shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java

```diff
@@ -0,0 +1,49 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.embedding.android;
+
+import android.app.Activity;
+import android.content.Context;
+import android.graphics.Rect;
+import android.view.Window;
+import android.view.WindowInsets;
+import androidx.annotation.RequiresApi;
+import androidx.annotation.VisibleForTesting;
+import io.flutter.Build;
+import io.flutter.util.ViewUtils;
+import java.util.Collections;
+import java.util.List;
+
+/** A delegate class that performs the task of retrieving the bounding rect values. */
+public class FlutterViewDelegate {
+  /**
+   * Return the WindowInsets object for the provided Context. Returns null if there is no associated
+   * activity, the window of the associated activity is null, or the root window insets of the
+   * activity's window is null.
+   */
+  @RequiresApi(api = Build.API_LEVELS.API_23)
+  @VisibleForTesting
+  public WindowInsets getWindowInsets(Context context) {
+    Activity activity = ViewUtils.getActivity(context);
+    if (activity == null) {
+      return null;
+    }
+    Window window = activity.getWindow();
+    if (window == null) {
+      return null;
+    }
+    return window.getDecorView().getRootWindowInsets();
+  }
+
+  @RequiresApi(api = Build.API_LEVELS.API_35)
+  @VisibleForTesting
+  public List<Rect> getCaptionBarInsets(Context context) {
+    WindowInsets insets = getWindowInsets(context);
```

Comment:
It looks like we should be moving to window insets for all android code after api 30. Not in this pr but in general. Based on your understanding of this android code is that correct?

Reply:
I don't understand your question. What do you mean by "moving to window insets"? Moving what, and from where?

Reply:
Let me try to rephrase: 
Do you, as the person who most recently read and worked with the window insets and its documentation, think that on api 30+ we need to rewrite any of the code that sets ViewportMetrics to ensure that ViewportMetrics is getting the correct values from the android embedder?

Reply:
The most comprehensive response I can give is that we currently already use the `WindowInsets` API, but the older inset getter method rather than the newer bounding rects method. Switching to the newer method in the future would allow us to cover more robust scenarios, such as where the safe area of a view is not a simple rectangle, but this would also require updating the implementation of `SafeArea`, and it looks to me like that update would constitute a breaking change.

Reply:
Ok so either a breaking change to safe area or a different new widget that let items grow in irregular shapes. Thank you, for now I think this means we dont need a rewrite and I appreciate the answer.